connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SessionHolder.scala

```diff
@@ -205,14 +220,26 @@ case class SessionHolder(userId: String, sessionId: String, session: SparkSessio
    * Called only by SparkConnectSessionManager.
    */
   private[connect] def close(): Unit = {
+    // It is called only by SparkConnectSessionManager.closeSession, only after it's removed from
+    // the sessionStore there guarantees that it is called only once.
+    if (closedTime.isDefined) {
+      throw new IllegalStateException(s"Session $key is already closed.")
+    }
+
     logInfo(s"Closing session with userId: $userId and sessionId: $sessionId")
 
-    // After isClosing=true, SessionHolder.addExecuteHolder() will not allow new executions for
-    // this session. Because both SessionHolder.addExecuteHolder() and
+    // After closedTime is defined, SessionHolder.addExecuteHolder() will not allow new executions
+    // to be added for this session. Because both SessionHolder.addExecuteHolder() and
     // SparkConnectExecutionManager.removeAllExecutionsForSession() are executed under
     // executionsLock, this guarantees that removeAllExecutionsForSession triggered below will
     // remove all executions and no new executions will be added in the meanwhile.
-    isClosing = true
+    closedTime = Some(System.currentTimeMillis())
+
+    if (eventManager.status == SessionStatus.Pending) {
+      // Testing-only: Some sessions created by SessionHolder.forTesting are not fully initialized
+      // and can't be closed.
+      return
+    }
```

Comment:
This is weird. I had was hitting issues from SparkConnectSessionManager.invalidateAllSessions because of some of these SessionHolder.forTesting session getting closed while not initialized (e.g. with null SparkSession and then hitting NPE trying to access artifactManager.

But it should have happened before when invalidateAllSessions() called sessionStore.invalidateAll(), and this should have triggered RemoveSessionListener.
The fact that tests were not failing suggests that either invalidateAll() did not trigger RemoveSessionListener, or that it swallowed errors from it. The latter may be possible, given that removal in Guava often happens via work-stealing piggy backed to other operations, so it would be prudent to eat all errors there. Now as we do it "ourselves", it surfaced.